homeassistant/components/recorder/__init__.py

```diff
@@ -314,127 +352,198 @@ def do_adhoc_purge(self, **kwargs):
+    @callback
+    def async_connection_failed(self):
+        """Connect failed tasks."""
+        self.async_db_ready.set_result(False)
+        persistent_notification.async_create(
+            self.hass,
+            "The recorder could not start, check [the logs](/config/logs)",
+            "Recorder",
+        )
+        self._stop_queue_watcher_and_event_listener()
 
-            if self.hass.state == CoreState.running:
-                hass_started.set_result(None)
-            else:
+    @callback
+    def async_connection_success(self):
+        """Connect success tasks."""
+        self.async_db_ready.set_result(True)
 
-                @callback
-                def notify_hass_started(event):
-                    """Notify that hass has started."""
-                    hass_started.set_result(None)
+    @callback
+    def _async_recorder_ready(self):
+        """Mark recorder ready."""
+        self.async_recorder_ready.set()
 
-                self.hass.bus.async_listen_once(
-                    EVENT_HOMEASSISTANT_START, notify_hass_started
-                )
+    @callback
+    def async_purge(self, now):
+        """Trigger the purge."""
+        self.queue.put(PurgeTask(self.keep_days, repack=False, apply_filter=False))
 
-        self.hass.add_job(register)
-        result = hass_started.result()
+    def run(self):
+        """Start processing events to save."""
+        shutdown_task = object()
+        hass_started = concurrent.futures.Future()
+
+        self.hass.add_job(
+            callback(lambda: self.async_register(shutdown_task, hass_started))
+        )
+
+        current_version = self._setup_recorder()
+
+        if current_version is None:
+            self.hass.add_job(self.async_connection_failed)
+            return
+
+        schema_is_current = migration.schema_is_current(current_version)
+        if schema_is_current:
+            self._setup_run()
+
+        self.hass.add_job(self.async_connection_success)
 
         # If shutdown happened before Home Assistant finished starting
-        if result is shutdown_task:
+        if hass_started.result() is shutdown_task:
             # Make sure we cleanly close the run if
             # we restart before startup finishes
             self._shutdown()
             return
 
-        # Start periodic purge
-        if self.auto_purge:
-
-            @callback
-            def async_purge(now):
-                """Trigger the purge."""
-                self.queue.put(
-                    PurgeTask(self.keep_days, repack=False, apply_filter=False)
+        # We wait to start the migration until startup has finished
+        # since it can be cpu intensive and we do not want it to compete
+        # with startup which is also cpu intensive
+        if not schema_is_current:
+            if self._migrate_schema_and_setup_run(current_version):
+                if not self._event_listener:
```

Comment:
Let's add a comment here that this can happen if we reached 30k pending events while migrating.

Reply:
done in 26582eb4f9